magicgui/_magicgui.py

```diff
@@ -1,18 +1,70 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Callable, Optional, Union, overload
+import inspect
+from functools import partial
+from types import FunctionType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generic,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
 from warnings import warn
 
 from typing_extensions import Literal
 
+from magicgui.widgets import FunctionGui, MainFunctionGui
+
 if TYPE_CHECKING:
     from magicgui.application import AppRef
-    from magicgui.widgets import FunctionGui, MainFunctionGui
+
+_T = TypeVar("_T")
+_R = TypeVar("_R")
+
+
+def _magicgui(function=None, factory=False, main_window=False, **kwargs):
+    """Actual private magicui decorator.
+
+    if factory is `True` will return a MagicFactory instance, that can be called
+    to return a `FunctionGui` instance.  See docstring of ``magicgui`` for parameters
+    """
+    if "result" in kwargs["param_options"]:
+        warn(
+            "\n\nThe 'result' option is deprecated and will be removed in the future."
+            "Please use `result_widget=True` instead.\n",
+            FutureWarning,
+        )
+
+        kwargs["param_options"].pop("result")
+        kwargs["result_widget"] = True
+
+    def inner_func(func: Callable) -> Union[FunctionGui, MagicFactory]:
+        if not callable(func):
+            raise TypeError("the first argument must be callable")
+
+        magic_class = MainFunctionGui if main_window else FunctionGui
+
+        if factory:
+            return MagicFactory(func, magic_class=magic_class, **kwargs)
+        return magic_class(func, **kwargs)
+
+    if function is None:
+        return inner_func
+    else:
+        return inner_func(function)
```

Comment:
In what situations does `function=None` make sense? Is this that weird trick that could be avoided with toolz.curry? =)

Reply:
yeah, we've had this discussion before! 😂  ... but I don't want to depend on toolz

Reply:
... and `function=None` makes sense when you provide parameters to the decorator:
```python
@magicgui(call_button=True)  # magicgui is being called with function=None
def function():
    ...
```

Reply:
Fine. 🤷 I think the toolz logic ends up so much cleaner, but I'll accept this for now. 😂

Reply:
i agree.  looks much better that way.